File: src/main/java/org/cubexmc/model/RedeemRecipe.java
package org.cubexmc.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * A complete set of preconditions that can redeem a gem.
 */
public final class RedeemRecipe {
    private final List<RedeemIngredient> requiresHeld;
    private final List<RedeemIngredient> consumes;
    private final List<RedeemIngredient> requiresRedeemed;
    private final List<String> requiresAny;
    private final int requiresCount;
    private final List<String> requiresCountFrom;

    public RedeemRecipe(List<RedeemIngredient> requiresHeld,
            List<RedeemIngredient> consumes,
            List<RedeemIngredient> requiresRedeemed,
            List<String> requiresAny,
            int requiresCount,
            List<String> requiresCountFrom) {
        this.requiresHeld = immutableIngredients(requiresHeld);
        this.consumes = immutableIngredients(consumes);
        this.requiresRedeemed = immutableIngredients(requiresRedeemed);
        this.requiresAny = immutableStrings(requiresAny);
        this.requiresCount = Math.max(0, requiresCount);
        this.requiresCountFrom = immutableStrings(requiresCountFrom);
    }

    public static RedeemRecipe legacy(List<String> requiresHeld,
            List<String> requiresRedeemed,
            List<String> consumes,
            List<String> requiresAny,
            int requiresCount,
            List<String> requiresCountFrom) {
        return new RedeemRecipe(toIngredients(requiresHeld), toIngredients(consumes), toIngredients(requiresRedeemed),
                requiresAny, requiresCount, requiresCountFrom);
    }

    private static List<RedeemIngredient> toIngredients(List<String> keys) {
        if (keys == null || keys.isEmpty()) {
            return Collections.emptyList();
        }
        List<RedeemIngredient> ingredients = new ArrayList<>();
        for (String key : keys) {
            if (key != null && !key.trim().isEmpty()) {
                ingredients.add(new RedeemIngredient(key, 1));
            }
        }
        return ingredients;
    }

    private static List<RedeemIngredient> immutableIngredients(List<RedeemIngredient> values) {
        if (values == null || values.isEmpty()) {
            return Collections.emptyList();
        }
        return Collections.unmodifiableList(new ArrayList<>(values));
    }

    private static List<String> immutableStrings(List<String> values) {
        if (values == null || values.isEmpty()) {
            return Collections.emptyList();
        }
        List<String> copy = new ArrayList<>();
        for (String value : values) {
            if (value != null && !value.trim().isEmpty()) {
                copy.add(value.trim());
            }
        }
        return copy.isEmpty() ? Collections.emptyList() : Collections.unmodifiableList(copy);
    }

    public boolean hasRequirements() {
        return !requiresHeld.isEmpty()
                || !consumes.isEmpty()
                || !requiresRedeemed.isEmpty()
                || !requiresAny.isEmpty()
                || (requiresCount > 0 && !requiresCountFrom.isEmpty());
    }

    public List<RedeemIngredient> getRequiresHeld() {
        return requiresHeld;
    }

    public List<RedeemIngredient> getConsumes() {
        return consumes;
    }

    public List<RedeemIngredient> getRequiresRedeemed() {
        return requiresRedeemed;
    }

    public List<String> getRequiresAny() {
        return requiresAny;
    }

    public int getRequiresCount() {
        return requiresCount;
    }

    public List<String> getRequiresCountFrom() {
        return requiresCountFrom;
    }
}

File: src/main/java/org/cubexmc/model/RedeemRecipe.kt
package org.cubexmc.model

import java.util.Collections
import kotlin.math.max

/**
 * A complete set of preconditions that can redeem a gem.
 */
class RedeemRecipe(
    requiresHeld: List<RedeemIngredient>?,
    consumes: List<RedeemIngredient>?,
    requiresRedeemed: List<RedeemIngredient>?,
    requiresAny: List<String>?,
    requiresCount: Int,
    requiresCountFrom: List<String>?,
) {
    val requiresHeld: List<RedeemIngredient> = immutableIngredients(requiresHeld)
    val consumes: List<RedeemIngredient> = immutableIngredients(consumes)
    val requiresRedeemed: List<RedeemIngredient> = immutableIngredients(requiresRedeemed)
    val requiresAny: List<String> = immutableStrings(requiresAny)
    val requiresCount: Int = max(0, requiresCount)
    val requiresCountFrom: List<String> = immutableStrings(requiresCountFrom)

    fun hasRequirements(): Boolean =
        requiresHeld.isNotEmpty() ||
            consumes.isNotEmpty() ||
            requiresRedeemed.isNotEmpty() ||
            requiresAny.isNotEmpty() ||
            (requiresCount > 0 && requiresCountFrom.isNotEmpty())

    companion object {
        @JvmStatic
        fun legacy(
            requiresHeld: List<String>?,
            requiresRedeemed: List<String>?,
            consumes: List<String>?,
            requiresAny: List<String>?,
            requiresCount: Int,
            requiresCountFrom: List<String>?,
        ): RedeemRecipe =
            RedeemRecipe(toIngredients(requiresHeld), toIngredients(consumes), toIngredients(requiresRedeemed), requiresAny, requiresCount, requiresCountFrom)

        private fun toIngredients(keys: List<String>?): List<RedeemIngredient> {
            if (keys.isNullOrEmpty()) return emptyList()
            val ingredients = ArrayList<RedeemIngredient>()
            for (key in keys) {
                if (!key.isNullOrBlank()) ingredients.add(RedeemIngredient(key, 1))
            }
            return ingredients
        }

        private fun immutableIngredients(values: List<RedeemIngredient>?): List<RedeemIngredient> {
            if (values.isNullOrEmpty()) return emptyList()
            return Collections.unmodifiableList(ArrayList(values))
        }

        private fun immutableStrings(values: List<String>?): List<String> {
            if (values.isNullOrEmpty()) return emptyList()
            val copy = ArrayList<String>()
            for (value in values) {
                if (!value.isNullOrBlank()) copy.add(value.trim())
            }
            return if (copy.isEmpty()) emptyList() else Collections.unmodifiableList(copy)
        }
    }
}
